src/libstd/sys/unix/net.rs

```diff
@@ -66,10 +67,31 @@ impl Socket {
     }
 
     pub fn duplicate(&self) -> io::Result<Socket> {
-        let fd = try!(cvt(unsafe { libc::dup(self.0.raw()) }));
-        let fd = FileDesc::new(fd);
-        fd.set_cloexec();
-        Ok(Socket(fd))
+        use libc::funcs::posix88::fcntl::fcntl;
+        const ZERO: c_int = 0;
+        static EMULATE_F_DUPFD_CLOEXEC: AtomicBool = AtomicBool::new(false);
+        if !EMULATE_F_DUPFD_CLOEXEC.load(atomic::Ordering::Relaxed) {
+            match cvt(unsafe { fcntl(self.0.raw(), libc::F_DUPFD_CLOEXEC, ZERO) }) {
+                // `EINVAL` can only be returned on two occasions: Invalid
+                // command (second parameter) or invalid third parameter. 0 is
+                // always a valid third parameter, so it must be the second
+                // parameter.
+                //
+                // Store the result in a global variable so we don't try each
+                // syscall twice.
+                Err(ref e) if e.raw_os_error() == Some(libc::EINVAL) => {
+                    EMULATE_F_DUPFD_CLOEXEC.store(true, atomic::Ordering::Relaxed);
+                }
+                res => return res.map(|fd| Socket(FileDesc::new(fd))),
```

Comment:
An interesting point of the MUSL thread indicated that some newer kernels apparently broke `F_DUPFD_CLOEXEC` such that it was the same as `F_DUPFD`, and hence even if this syscall succeeds [they still set cloexec](https://github.com/wermut/musl/blob/55ce8300293ffe5a22982a8fd35de11f68eae3af/src/fcntl/fcntl.c#L27-L28). It may be good to just keep the same common code at the end where `set_cloexec` is called?